File: problems/310_minimum-height-trees/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>> &edges)
    {
        if (n == 0 || n == 1)
            return {0};
        vector<int> indegree(n, 0);
        vector<vector<int>> graph(n);
        for (auto e : edges)
        {
            graph[e[0]].push_back(e[1]);
            graph[e[1]].push_back(e[0]);
            indegree[e[1]]++;
            indegree[e[0]]++;
        }
        queue<int> q;
        for (int i = 0; i < n; i++)
            if (indegree[i] == 1)
                q.push(i); // push all the root nodes possible
        int remaining = n;
        while (remaining > 2)
        { // Removing leaves layer by layer
            int sz = (int)q.size();
            remaining -= sz;

            while (sz--)
            {
                int leaf = q.front();
                q.pop();
                for (int nei : graph[leaf])
                {
                    if (--indegree[nei] == 1)
                        q.push(nei);
                }
            }
        }

        vector<int> res;
        while (!q.empty())
        {
            res.push_back(q.front());
            q.pop();
        }
        return res;
    }
};

static void printVector(const vector<int> &values)
{
    cout << "[";
    for (size_t i = 0; i < values.size(); ++i)
    {
        if (i > 0)
        {
            cout << ",";
        }
        cout << values[i];
    }
    cout << "]";
}

int main()
{
    Solution solution;

    vector<vector<int>> edges1 = {{1, 0}, {1, 2}, {1, 3}};
    vector<int> result1 = solution.findMinHeightTrees(4, edges1);
    cout << "Sample 1 placeholder result: ";
    printVector(result1);
    cout << "\nSample 1 expected output: [1]\n";

    vector<vector<int>> edges2 = {{3, 0}, {3, 1}, {3, 2}, {3, 4}, {5, 4}};
    vector<int> result2 = solution.findMinHeightTrees(6, edges2);
    cout << "Sample 2 placeholder result: ";
    printVector(result2);
    cout << "\nSample 2 expected output: [3,4]\n";

    return 0;
}

File: problems/310_minimum-height-trees/main.rs
struct Solution;

impl Solution {
    pub fn find_min_height_trees(n: i32, edges: Vec<Vec<i32>>) -> Vec<i32> {
        let n = n as usize;
        if n <= 1 {
            return vec![0];
        }
        let mut graph: Vec<Vec<usize>> = vec![vec![]; n];
        let mut indegree: Vec<i32> = vec![0; n];
        for edge in edges {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            graph[u].push(v);
            graph[v].push(u);
            indegree[u] += 1;
            indegree[v] += 1;
        }
        use std::collections::VecDeque;
        let mut queue = VecDeque::new();
        for i in 0..n {
            if indegree[i] == 1 {
                queue.push_back(i as i32);
            }
        }
        let mut remaining = n;
        while remaining > 2 {
            let sz = queue.len();
            remaining -= sz;
            for _ in 0..sz {
                let leaf = queue.pop_front().unwrap() as usize;
                for &nei in &graph[leaf] {
                    indegree[nei] -= 1;
                    if indegree[nei] == 1 {
                        queue.push_back(nei as i32);
                    }
                }
            }
        }
        let mut res = Vec::new();
        while let Some(x) = queue.pop_front() {
            res.push(x);
        }
        res
    }
}

fn print_vec(values: &[i32]) {
    print!("[");
    for (idx, value) in values.iter().enumerate() {
        if idx > 0 {
            print!(",");
        }
        print!("{value}");
    }
    println!("]");
}

fn main() {
    let edges1 = vec![vec![1, 0], vec![1, 2], vec![1, 3]];
    let result1 = Solution::find_min_height_trees(4, edges1);
    print!("Sample 1 placeholder result: ");
    print_vec(&result1);
    println!("Sample 1 expected output: [1]");

    let edges2 = vec![vec![3, 0], vec![3, 1], vec![3, 2], vec![3, 4], vec![5, 4]];
    let result2 = Solution::find_min_height_trees(6, edges2);
    print!("Sample 2 placeholder result: ");
    print_vec(&result2);
    println!("Sample 2 expected output: [3,4]");
}
